beginner_source/hybrid_frontend/introduction_to_hybrid_frontend_tutorial.py

```diff
@@ -0,0 +1,526 @@
+    test_len = int(len(dataset) - train_len)
+    # Split train and test sets
+    train_set = dataset[:train_len]
+    test_set = dataset[-test_len:]
+    # Convert splits to torch tensor and return
+    return torch.tensor(train_set), torch.tensor(test_set)
+
+
+# Split dataset
+train_set, test_set = split_dataset(dataset, train_ratio)
+
+
+######################################################################
+# Implement k-NN functions
+# ~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# The next step is to implement the helper functions that we will use to
+# carry out the k-NN algorithm. Because we use torch tensors to operate
+# on, we can use the PyTorch JIT compiler to compile our functions to a
+# graph representation.
+#
+# getDistance
+# ^^^^^^^^^^^
+#
+# The first function that we'll implement is ``getDistance``, which is how
+# we will calculate the distance between a ``test_instance`` and a
+# ``train_instance``. Because our features are all real numbers, we will
+# use Euclidean distance:
+#
+# .. math:: distance = \sqrt{(train[0] - test[0])^2 + (train[1] - test[1])^2 + \ldots +  (train[n] - test[n])^2}
+#
+# --------------
+#
+# **Trace mode**
+#
+# Because this function involves sequential torch tensor operations, with
+# no data-dependent control flows, we can use the **trace** function to
+# compile it to a graph representation.
+#
+# Notice that to trace a function, you must first declare the traced
+# version of the function (``getDistance_traced``) using the
+# ``torch.jit.trace`` function. Notice that the ``trace`` function is a
+# `Curried function <https://en.wikipedia.org/wiki/Currying>`__. For us
+# this means that calling it requires two arguments, each surrounded by
+# its own set of parenthesis. In the example below, we show that the
+# outputs of the non-compiled and compiled functions are identical, and
+# use the ``.graph`` attribute to show the graph in a human readable
+# format.
+#
+
+# Takes a train_instance and test_instance, and returns a scalar torch
+#  tensor containing the distance value
+def getDistance(train_instance, test_instance):
+    # Element wise subtraction; disregard last element (class)
+    #diff = torch.sub(train_instance[:4], test_instance[:4])
+    diff = train_instance[:4] - test_instance[:4]
+    # Square each element of the distance tensor
+    #sqrs = torch.pow(diff, torch.tensor(2.0))
+    sqrs = diff ** torch.tensor(2.0)
+    # Sum all elements in sqrs tensor
+    sumsqrs = torch.sum(sqrs)
+    # Take square root of sumsqrs to obtain distance
+    distance = torch.sqrt(sumsqrs)
+    return distance
+
+
+########## Example ##########
+# Non-compiled
+print("Euclidean distance between train_set[0] and train_set[1]:")
+print("train_set[0]:", train_set[0])
+print("train_set[1]:", train_set[1])
+distance = getDistance(train_set[0], train_set[1])
+print("Non-compiled output:", distance)
+
+# Compiled
+getDistance_traced = trace(train_set[0], train_set[1])(getDistance)
```

Comment:
we updated the API to look nicer here:

`trace(getDistance, (train_set[0], train_set[1]))`